pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,474 @@
+
+use cfg_primitives::{SECONDS_PER_MONTH, SECONDS_PER_YEAR};
+use cfg_traits::{interest::InterestRate, Seconds};
+use chrono::{DateTime, Datelike, Days, Months, NaiveDate, TimeDelta};
+use chronoutil::DateRule;
+use frame_support::pallet_prelude::RuntimeDebug;
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{
+		EnsureAddAssign, EnsureDiv, EnsureFixedPointNumber, EnsureInto, EnsureSub, EnsureSubAssign,
+	},
+	ArithmeticError, DispatchError, FixedPointNumber, FixedPointOperand,
+};
+
+/// Specify the expected repayments date
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum Maturity {
+	/// Fixed point in time, in secs
+	Fixed {
+		/// Secs when maturity ends
+		date: Seconds,
+		/// Extension in secs, without special permissions
+		extension: Seconds,
+	},
+}
+
+impl Maturity {
+	pub fn fixed(date: Seconds) -> Self {
+		Self::Fixed { date, extension: 0 }
+	}
+
+	pub fn date(&self) -> Seconds {
+		match self {
+			Maturity::Fixed { date, .. } => *date,
+		}
+	}
+
+	pub fn is_valid(&self, now: Seconds) -> bool {
+		match self {
+			Maturity::Fixed { date, .. } => *date > now,
+		}
+	}
+
+	pub fn extends(&mut self, value: Seconds) -> Result<(), ArithmeticError> {
+		match self {
+			Maturity::Fixed { date, extension } => {
+				date.ensure_add_assign(value)?;
+				extension.ensure_sub_assign(value)
+			}
+		}
+	}
+}
+
+fn seconds_to_days(seconds: Seconds) -> Result<u32, DispatchError> {
+	Ok(TimeDelta::try_seconds(seconds.ensure_into()?)
+		.ok_or(DispatchError::Other("Precission error with seconds"))?
+		.num_days()
+		.ensure_into()?)
+}
+
+fn seconds_to_date(date_in_seconds: Seconds) -> Result<NaiveDate, DispatchError> {
+	Ok(DateTime::from_timestamp(date_in_seconds.ensure_into()?, 0)
+		.ok_or(DispatchError::Other("Invalid date in seconds"))?
+		.date_naive())
+}
+
+fn date_to_seconds(date: NaiveDate) -> Result<Seconds, DispatchError> {
+	Ok(date
+		.and_hms_opt(0, 0, 0)
+		.ok_or(DispatchError::Other("Invalid h/m/s"))?
+		.and_utc()
+		.timestamp()
+		.ensure_into()?)
+}
+
+fn seconds_in_month(date: NaiveDate) -> Result<Seconds, DispatchError> {
+	todo!()
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum ReferenceDate {
+	/// Payments are expected every period relative to a specific date.
+	/// E.g. if the period is monthly and the specific date is Mar 3, the
+	/// first interest payment is expected on Apr 3.
+	Date(Seconds),
+
+	/// At the end of the period, e.g. the last day of the month for a monthly
+	/// period
+	End,
+}
+
+impl ReferenceDate {
+	fn monthly_cashflow_dates(
+		&self,
+		start: Seconds,
+		end: Seconds,
+	) -> Result<Vec<(Seconds, Seconds)>, DispatchError> {
+		if start > end {
+			return Err(DispatchError::Other("cashflow must start before it ends"));
+		}
+
+		let start = seconds_to_date(start)?;
+		let end = seconds_to_date(end - 1)? + Months::new(1);
```

Comment:
WHy the minus 1?

Reply:
We want as a final date for iterating months at least one month less a second.

It means, if your loan is valid until the first second of March 1, then you can pay until March 31, 23:59:59:999. That's the theoretical worst case.

Reply:
Where does this requirement come from? Why can't you pay on that date inclusive? I would assume it is inclusive. Cc @denniswell

Reply:
If you need to pay on the last day of the month, and your overdue date is on day 1 in a month, then you can pay until that month ends.

That was my conclusion from: https://centrifugedao.slack.com/archives/C05DX0RH1T3/p1712920193116179

Reply:
Ok but make sense too to pay the same day of maturity 👍🏻